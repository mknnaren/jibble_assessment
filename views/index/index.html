    <div>
        <v-row justify="center" align="center">
            <v-col cols="12" >
                <v-card>
                    <v-card-title class="headline">
                        Movie List <v-btn small color="primary" @click="getMovieList">
                            Refresh
                        </v-btn>
                    </v-card-title>
                    
                </v-card>
            </v-col>
        </v-row>
        <v-row justify="center" align="center">
            <v-col cols="12" sm="8" md="6">
                <div class="text-center">
                    <MovieTableSearch @show-search-list="getSearchResult"></MovieTableSearch>
                    <MovieTable :movieData = "movieData"></MovieTable>
                    <MovieTablePage @show-page-list="getPageMovieList" :pageData = "pageData"></MovieTablePage>
                </div>
            </v-col>
        </v-row>
    </div>