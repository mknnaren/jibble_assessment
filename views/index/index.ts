import axios from 'axios';
import { defineComponent, computed, onMounted, ref } from '@nuxtjs/composition-api'
import MovieTable from '~/components/movie_table.vue'
import MovieTablePage from '~/components/movie_table_page.vue'
import MovieTableSearch from '~/components/movie_table_search.vue'

export default defineComponent({
    components: {
        MovieTable,
        MovieTablePage,
        MovieTableSearch
    },
    props: {},

    setup( props , context) {
        const movTitle = ref("");
        const page = ref(1);
        const pageData = ref({});
        const movieData = ref({
            loading: true,
            data: [],
            total: 0
        });
        function getPageMovieList(pageNo: number) {
            page.value = pageNo;
            getMovieList();
        }
        function getSearchResult(searchTxt: string) {
            movTitle.value = searchTxt;
            page.value = 1;
            getMovieList();
        }
        
        function getMovieList() {
            movieData.value.loading = true;
            axios({
                method: 'get',
                baseURL: "https://jsonmock.hackerrank.com/api/movies/search",
                url: `/?Title=${movTitle.value}&page=${page.value}`,
            })
                .then(response => {
                    console.log(response.data);
                    response.data.page = Number(response.data.page);
                    for (let i = 0; i < response.data.data.length; i++){
                        response.data.data[i]["favourite"] = true;
                    }
                    movieData.value = response.data;
                    movieData.value.loading = false;
                    const { data, ...pageObj } = response.data
                    pageData.value = pageObj;
                    console.log(pageData.value);
                }, err => {
                    console.log(err);
                });
        }

        onMounted(() => {
            getMovieList();
        });
        
        // const filteredMovieList = computed(function () {
        //     return movieLs.value.filter(
        //         (movieLs) => movieLs.title.includes(movTitle)
        //     );
        // });
        return {
            getMovieList,
            getPageMovieList,
            getSearchResult,
            movieData,
            pageData
        }
    }
})

