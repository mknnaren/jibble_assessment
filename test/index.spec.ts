import { mount, shallowMount, createLocalVue  } from '@vue/test-utils'
import index from '@/pages/index.vue';
import searchComp from '@/components/movie/search.vue';
import tableComp from '@/components/movie/table.vue';
import addfavComp from '@/components/movie/add_to_fav.vue';
import pageComp from '@/components/movie/page.vue';
import Vuetify from 'vuetify';
import Vue from 'vue'
import VueRouter from 'vue-router'
Vue.use(Vuetify)

describe('index', () => {
  let localVue = createLocalVue()
  let router = new VueRouter()
  let vuetify = new Vuetify()
  test('mounts properly', () => {
    const wrapper = mount(index, {localVue, router, vuetify,  propsData: {
      movListProp: [{
        imdbID: "ab123", Title: "Harry", favourite: true, Year: "2021" 
      }],
      movPageProp: {
        page: 1,
        per_page: 10,
        total_pages: 10,
        total: 100
      },
      loadingProp: true
    }})
    expect(wrapper.vm).toBeTruthy()
  })

  test('mounts child component SEARCH properly', () => {
    const wrapper = mount(index, {localVue, router, vuetify,  propsData: {
      movListProp: [{
        imdbID: "ab123", Title: "Harry", favourite: true, Year: "2021" 
      }],
      movPageProp: {
        page: 1,
        per_page: 10,
        total_pages: 10,
        total: 100
      },
      loadingProp: true
    }})
    const searchElem = wrapper.findComponent(searchComp) // => finds Bar by component instance
    expect(searchElem.exists()).toBe(true)
  })
  test('mounts child component TABLE properly', () => {
    const wrapper = mount(index, {localVue, router, vuetify,  propsData: {
      movListProp: [{
        imdbID: "ab123", Title: "Harry", favourite: true, Year: "2021" 
      }],
      movPageProp: {
        page: 1,
        per_page: 10,
        total_pages: 10,
        total: 100
      },
      loadingProp: true
    }})
    const tableElem = wrapper.findComponent(tableComp) // => finds Bar by component instance
    expect(tableElem.exists()).toBe(true)
  })
  test('mounts child component PAGINATION properly', () => {
    const wrapper = mount(index, {localVue, router, vuetify,  propsData: {
      movListProp: [{
        imdbID: "ab123", Title: "Harry", favourite: true, Year: "2021" 
      }],
      movPageProp: {
        page: 1,
        per_page: 10,
        total_pages: 10,
        total: 100
      },
      loadingProp: true
    }})
    const pageElem = wrapper.findComponent(pageComp) // => finds Bar by component instance
    expect(pageElem.exists()).toBe(true)
  })
  
  test('check for search emit event propagation from child component', async() => {
    const wrapper = mount(index, {localVue, vuetify, router,  propsData: {
      movListProp: [{
        imdbID: "ab123", Title: "Harry", favourite: true, Year: "2021" 
      }],
      movPageProp: {
        page: 1,
        per_page: 10,
        total_pages: 10,
        total: 100
      },
      loadingProp: true
    }})
    const searchElem = wrapper.findComponent(searchComp) // => finds Bar by component instance
    expect(searchElem.exists()).toBe(true)
    const input = searchElem.find('input')
    input.trigger('focus')
    // input.element.value = 'h'
    input.trigger('input')
    input.trigger('keydown.enter')
    input.trigger('keydown.enter')
    // expect(wrapper.emitted()["show-search-list"]).toBeTruthy()
  })
  test('check for update-table event propagation from child component', async() => {
    const wrapper = mount(index, {localVue, vuetify, router, propsData: {
      movListProp: [{
        imdbID: "ab123", Title: "Harry", favourite: true, Year: "2021" 
      }],
      movPageProp: {
        page: 1,
        per_page: 10,
        total_pages: 10,
        total: 100
      },
      loadingProp: true
    }})
    const tableElem = wrapper.findComponent(tableComp) // => finds Bar by component instance
    expect(tableElem.exists()).toBe(true)
    const favStarElem = tableElem.findComponent(addfavComp)
    expect(favStarElem.exists()).toBe(true)
    await favStarElem.find('div#favStarDiv').trigger('click')
  })

  test('renders properly', () => {
    const wrapper = mount(index,  {localVue, router, vuetify})
    expect(wrapper.html()).toMatchSnapshot()
  })
})
