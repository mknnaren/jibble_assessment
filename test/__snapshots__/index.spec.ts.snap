// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index renders properly 1`] = `
"<div>
  <div class=\\"row align-center justify-center\\">
    <div class=\\"col-sm-8 col-md-6 col-12\\">
      <div class=\\"v-card v-sheet theme--light\\">
        <div class=\\"v-card__title headline\\">
          Movie List
          <div class=\\"spacer\\"></div> <button type=\\"button\\" class=\\"v-btn v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--round theme--light v-size--small primary\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-refresh theme--light\\"></i></span></button></div>
      </div>
    </div>
  </div>
  <div class=\\"row align-center justify-center\\">
    <div class=\\"col-sm-8 col-md-6 col-12\\">
      <div class=\\"text-center\\">
        <div>
          <div class=\\"text-center pt-2\\">
            <div class=\\"v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined\\">
              <div class=\\"v-input__control\\">
                <div class=\\"v-input__slot\\">
                  <div class=\\"v-input__prepend-inner\\">
                    <div class=\\"v-input__icon v-input__icon--prepend-inner\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-magnify theme--light\\"></i></div>
                  </div>
                  <fieldset aria-hidden=\\"true\\">
                    <legend style=\\"width: 0px;\\"><span>â€‹</span></legend>
                  </fieldset>
                  <div class=\\"v-text-field__slot\\"><label for=\\"input-163\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: absolute;\\">Search</label><input id=\\"input-163\\" type=\\"text\\"></div>
                </div>
                <div class=\\"v-text-field__details\\">
                  <div class=\\"v-messages theme--light\\">
                    <transition-group-stub tag=\\"div\\" name=\\"message-transition\\" class=\\"v-messages__wrapper\\"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div class=\\"v-data-table elevation-1 theme--light\\">
            <div class=\\"v-data-table__wrapper\\">
              <table>
                <colgroup>
                  <col class=\\"\\">
                  <col class=\\"\\">
                  <col class=\\"\\">
                  <col class=\\"\\">
                </colgroup>
                <thead class=\\"v-data-table-header\\">
                  <tr>
                    <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"Title\\" class=\\"text-left\\"><span>Title</span></th>
                    <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"Year\\" class=\\"text-start\\"><span>Year</span></th>
                    <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"IMDB ID\\" class=\\"text-start\\"><span>IMDB ID</span></th>
                    <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"\\" class=\\"text-start\\"><span></span></th>
                  </tr>
                </thead>
                <thead>
                  <tr class=\\"v-data-table__progress\\">
                    <th colspan=\\"4\\" class=\\"column\\">
                      <div role=\\"progressbar\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\" class=\\"v-progress-linear v-progress-linear--absolute theme--light\\" style=\\"height: 4px;\\">
                        <div class=\\"v-progress-linear__background primary\\" style=\\"opacity: 0.3; left: 0%; width: 100%;\\"></div>
                        <div class=\\"v-progress-linear__buffer\\"></div>
                        <transition-stub name=\\"fade-transition\\">
                          <div class=\\"v-progress-linear__indeterminate v-progress-linear__indeterminate--active\\">
                            <div class=\\"v-progress-linear__indeterminate long primary\\"></div>
                            <div class=\\"v-progress-linear__indeterminate short primary\\"></div>
                          </div>
                        </transition-stub>
                      </div>
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr class=\\"v-data-table__empty-wrapper\\">
                    <td colspan=\\"4\\">Loading items...</td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <div>
          <div class=\\"text-center pt-2\\">
            <nav role=\\"navigation\\" aria-label=\\"Pagination Navigation\\" id=\\"pagiElem\\">
              <ul class=\\"v-pagination theme--light\\">
                <li><button disabled=\\"disabled\\" type=\\"button\\" aria-label=\\"Previous page\\" class=\\"v-pagination__navigation v-pagination__navigation--disabled\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-chevron-left theme--light\\"></i></button></li>
                <li><button disabled=\\"disabled\\" type=\\"button\\" aria-label=\\"Next page\\" class=\\"v-pagination__navigation v-pagination__navigation--disabled\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-chevron-right theme--light\\"></i></button></li>
              </ul>
            </nav>
            <div class=\\"v-input v-input--is-readonly theme--light v-text-field\\">
              <div class=\\"v-input__control\\">
                <div class=\\"v-input__slot\\">
                  <div class=\\"v-text-field__slot\\"><label for=\\"input-178\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: absolute;\\">Items per page</label><input id=\\"input-178\\" readonly=\\"readonly\\" type=\\"number\\"></div>
                </div>
                <div class=\\"v-text-field__details\\">
                  <div class=\\"v-messages theme--light\\">
                    <transition-group-stub tag=\\"div\\" name=\\"message-transition\\" class=\\"v-messages__wrapper\\"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
